sdk/identity/azure-identity/inc/azure/identity/managed_identity_credential.hpp

```diff
@@ -42,6 +42,20 @@ namespace Azure { namespace Identity {
     std::string ToString() const { return m_resourceId; }
   };
 
+  /**
+   * @brief Options for managed identity credential.
+   *
+   */
+  struct ManagedIdentityCredentialOptions final : public Core::Credentials::TokenCredentialOptions
+  {
+    /**
+     * @brief The object ID of the service principal used to authenticate a user-assigned managed
+     * identity. This is an alternative to providing a client ID, and isn't required for
+     * system-assigned managed identities.
+     */
+    std::string ObjectId;
```

Comment:
Why is objectID the only one in the options? Its a required parameter if you are creating managedIdentity with ObjectID.

Reply:
Is there a spec for the design?

Reply:
**None** of object id, client id, or resource id are required parameters to use managed identity.

Please take a look at the Imds endpoint documentation, for reference, which mentions that all these are **optional** query parameters:
https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/how-to-use-vm-token#get-a-token-using-http 

If no parameters are passed in, which is allowed, the system-assigned managed identity is used. That is what happens in the case of DAC (where no required parameters are passed in, just the `TokenCredentialOptions` bag).

Here's another reference doc for AppService:
https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#rest-endpoint-reference
> Cannot be used on a request that includes principal_id, mi_res_id, or object_id. If all ID parameters (client_id, principal_id, object_id, and mi_res_id) are omitted, the system-assigned identity is used.

In the UAI scenario, only one of the three IDs needs to get set. For SAI, none needs to be set.

Even the existing `ManagedIdentityCredential` ctor which accepts `clientId` has it defaulted (i.e. not required), so the user can choose to pass in a `clientId` if they want, but don't have to.

We can certainly choose to add `clientId` and `resourceId` in the options bag, in the future, in a non-breaking way. There's no urgent need to add those today. Though not required, we can also then consider either deprecating the existing ctor (to minimize having more than one way to do the same thing).

A downside of adding the other, effectively optional parameters, eagerly, is that now you'd have to validate the state of the input parameters and throw if the user provides both objectId and clientId (or have an opinionated preference of one over the other, which isn't the best approach).

Reply:
What I'm saying is, If you want a UA managedIdenity based on object_id you need to pass the optional object_id. That's why our API has it as required even though under the hood we are passing an "optional". We chose to make the signature that way.

We are choosing to add resource_id as mandatory. Though we could change that since it hasn't shipped yet.

I agree that we can't overload the signature to have object_id as mandatory because two of the three are strings. However I don't consider object_id to be any less mandatory than the other two signatures. Hence why call it optional? Why isn't this an object `ManagedIdentityConfiguration` or some other type of object taking all three fields, granted only allow 1 of the three to be set/passed.

Reply:
I think we are looking at the same design choices a bit differently. I preferred the current option, given the trade-offs I mentioned, but it is reasonable to pick a different approach weighing the alternatives.
 
> Why isn't this an object ManagedIdentityConfiguration or some other type of object taking all three fields,

Sure, we can. There are trade-offs here. **Is that what you prefer as the design?**

> granted only allow 1 of the three to be set/passed

That's why. One approach doesn't result in the user getting into a state that requires us checking/throw. The other requires validation to ensure mutual exclusion. If we are fine with that trade-off, we can certainly make that happen.

> We are choosing to add resource_id as mandatory

To be clear, given the overloads we currently have, resource_id is effectively **not** a mandatory parameter. It might appear as mandatory, but hopefully this example can clarify the effective outcome.

There are two ways to achieve "optional" parameters:
1) Have one ctor with defaulted argument: `Foo(int x = 0);`
2) Have two ctor overloads: `Foo();` and `Foo(int x);`

Both are effectively the same thing. With option 2, x is still considered an optional parameter, achieving that with overloading instead. I wouldn't call x as a mandatory parameter here.

Reply:
We should use required vs mandatory. The thing we did here is we promoted an optional parameter into the signature. Being outside the options bag is normally only done for required parameters. Required parameters thus become mandatory even though sometimes in C++ we put defaults on them.

Reply:
I don't understand the design feedback and intent here. We can use different definitions, if you like, but what's the relevance to this PR?
Semantics of definitions aside, would you prefer we go with a different solution that you like more, over what I have proposed, given the trade-offs I have mentioned?

If so, please suggest it concretely, and your rationale why the trade-off analysis I did doesn't make sense. If you like your solution more, I am fine with going with that approach to move this PR and design along.

Reply:
Let's schedule a meeting tomorrow with the key stakeholders (You, Larry, me) to finalize and close out on the PR.

Reply:
Based on offline discussion, we decided to go with the approach of using `Uuid` as the type for `objectId` to deal with the ctor overloads with string `clientId`.

This way, we don't end up needing a `ManagedIdentityCredentialOptions` accepting ctor overload.
To do this, we require adding a `Uuid::CreateFromString` factory method (which parses the uuid string).

Reply:
It seems strange to differentiate only one of the Id values as uuid just to satisfy the overload ambiguity. Since we haven't GA'd this yet, have we considered creating a special type to represent any of the Id values? MSAL did something similar [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/src/client/Microsoft.Identity.Client/AppConfig/ManagedIdentityId.cs).

Reply:
Yep. That approach is like a discriminated union (similar to what GoLang did), where we have a "type/kind" discriminator with the input value. C++ 17 has `std::variant` to express a union type like this (the SDK targets C++ 14). We decided that approach has a similar downside where the user can pass in mismatched value, so wasn't worth the trade-off to introduce a new type.

We considered a similar option (see https://github.com/Azure/azure-sdk-for-cpp/pull/5914/files#diff-af093c7b7fd475c58855bd67fde2cf497bf92cd9190d95ee53fb7bc43a5555e0R54), where we'd add all three Ids to the `ManagedCredentialOptions` bag (and remove the `resourceId` accepting ctor, which .NET can't do). That was a viable option (but we ranked it second) because that approach can cause the user to pass in multiple IDs (and hence could be in an invalid state).

Therefore, we decided there is value in keeping the simpler pattern of overloading. The pattern remains consistent: 3 ctors, each independently accepting clientId, objectId, and resourceId.